assets/blocks/course-outline/status-store.js

```diff
@@ -0,0 +1,226 @@
+import { select as selectData, registerStore } from '@wordpress/data';
+import { createReducerFromActionMap } from '../../shared/data/store-helpers';
+import { Status } from './status-control';
+import { select, controls } from '@wordpress/data-controls';
+
+const DEFAULT_STATE = {
+	completedLessons: new Set(),
+	totalLessonsCount: 0,
+};
+
+/**
+ * Status store actions.
+ */
+const actions = {
+	/**
+	 * Sets the status of a lesson.
+	 *
+	 * @param {string} lessonId The lesson id.
+	 * @param {string} status   The lesson status.
+	 *
+	 * @return {Object} The setLessonStatus action.
+	 */
+	setLessonStatus( lessonId, status ) {
+		return {
+			type: 'SET_LESSON_STATUS',
+			lessonId,
+			status,
+		};
+	},
+
+	/**
+	 * Sets the status of a module by updating the status of its lessons.
+	 *
+	 * @param {string} moduleId The module block id.
+	 * @param {string} status   The module status.
+	 *
+	 * @return {Object} Yields the lesson update actions.
+	 */
+	*setModuleStatus( moduleId, status ) {
+		const currentStatus = yield select(
```

Comment:
Hmm not sure about this one. I would prefer to keep the signature simpler unless there is a performance impact. I checked and the method call takes less than 1ms so there isn't much to gain if we change that.  Additionaly the status store is responsible of keeping track of the current module status so it doesn't make much sense to expose it as an argument.

Reply:
Thinking a little more about that, maybe we could move this action logic (ifs and loops) to the reducer instead of keeping this in the actions. WDYT? But if you prefer to keep it like it's currently, it's okay to me ;)

Reply:
I thought about this too but wouldn't this break the convention that reducers should be [pure](https://redux.js.org/basics/reducers#handling-actions)?

Reply:
It'd be pure, if I didn't miss anything, the only thing we should get in the action would be the `lessonIds`. The rest I think we have the information in the state, right? So we'd only compare the previous and the new status.

Reply:
I still think that this wouldn't make the code clearer as now the `setModuleStatus` action is composed by `setLessonStatus` actions which makes it clear how the module status is calculated. Adding a new reducer to update the `completedLessons` would just make it more complicated.

I switched to using `setLessonStatus` to make it more clear and removed the same status checked as it was impossible to get it anyway.